### ISSUE TITLE
Btrfk is unable to do backup

### ISSUE THREAD

**Issue by Nicop06** [digint/btrbk#225]
Hello,

I am having issues when doing btrfs snapshot. It seems that I have a corruption somewhere but I am unable to figure out where. Here is the error I have:

```
WARNING: [send/receive] (send=/btrfs/snapshot.20180406T2232, receive=/backup/) At subvol /btrfs/home/.snapshot/nicolas.20180406T2232
WARNING: [send/receive] (send=/btrfs/home/.snapshot/nicolas.20180406T2232, receive=/backup/home) At subvol nicolas.20180406T2232
ERROR: send ioctl failed with -2: No such file or directory
ERROR: short read from stream: expected 31095 read 26999                        
ERROR: Failed to send/receive btrfs subvolume: /btrfs/home/.snapshot/nicolas.20180406T2232  -> /backup/home
WARNING: Deleted partially received (garbled) subvolume: /backup/home/nicolas.20180406T2232
ERROR: Error while resuming backups, aborting          
```

`/btrfs` is my main disk (SSD) and `/backup` is a local HDD I use for backups.

I noticed this issue this issue as btrbk was not pruning the snapshots, causing a lot of space issues on my SSD. I think the issue appeared around February but I just noticed it. Here are some information about my setup, which were likely different at the time the issue appeared considering I am using a rolling release :

```
$ btrbk --version
btrbk command line client, version 0.26.1
$ btrfs --version
btrfs-progs v4.16.1
$ uname -r
4.16.6_1
$ lsb_release -a
LSB Version:	1.0
Distributor ID:	VoidLinux
Description:	Void Linux
Release:	rolling
Codename:	void
```

Here is the list of actions I took in order to fix the problem without success:
* `btrbk clean`
* `btrbk prune`
* `btrfs scrub`: no error found
* Deleting the parent snapshot to force using full backup instead of incremental backup. I kept older backups though, none of which have a corresponding snapshot on the original subvolume. I do not know the effect of doing that, and I can try to delete them if you think this will solve the issue.

The logs I added are after doing all of these actions.

Also, I am using `btrbk` for many subvolumes and only this one has issues. The only obvious difference with the failing subvolume subvolume is that I am doing 2 backups: 1 to a local disk and 1 via ssh. The backup via ssh has also been failing since February while it is using a different  `snapshot_name` in `btrbk.conf`.

I know this problem is certainly not due to `btrbk` itself but more to `btrfs-progs` or the kernel itself, but any help to investigate this issue would be greatly appreciated.

Thank you.

**Comment by digint** [comment]
It's related/duplicate of #91, #223 : upstream bug, nothing that can be fixed in btrbk.

The change of `snapshot_name` is most probably not causing the issue, the btrfs ERROR indicates that you have a file missing on the target subvolume.

I usually recommend to "upgrade to latest kernel", but as you already are on newest, you might want to post this on the btrfs mailing list for further investigation:

Run `btrbk -l debug`, and grep for `###` to see the btrfs commands causing the trouble.

**Comment by Nicop06** [comment]
Thank you for the tip. Here is the `btrfs` command line causing the issue:

```
btrfs send /btrfs/home/.snapshot/nicolas.20180406T2232 | btrfs receive /backup/home/
```

In my case I am unsure that this is related to those problems as:

* I have the latest kernel (I even tried `4.17` without success as I upgraded today), and even in February I was already running 4.14 or 4.15.
* I do not have any snapshot parent (or at least removing it does not solve the issue).

It is definitely a `btrfs` issue so I will try to investigate it with btrfs developers. I just hoped that a btrbk user had the issue and would have already some element of answer. I will leave this issue open though to post my findings in case some other people are interested.

**Comment by gkzsolt** [comment]
Hi,
Can you tell please where to report these upstream bugs? I tried to subscribe to their [mailing list](https://btrfs.wiki.kernel.org/index.php/Problem_FAQ#How_do_I_report_bugs_and_issues.3F), but this seems to be only for submitting patches. So I'll report to bugzilla on kernel.org, and then pray for them to pick it up.
I installed BTRFS a few months ago to my machines, after doing some research and being impressed by its power. I was planning to have it on all machines and doing super powerful incremental backups using send/receive. Yesterday came the shock, when I realised that incremental backups just do NOT work!

I am doing the worlds simplest backup:
```
btrfs subvol snapshot -r / /snapshots/@root-12.05.2018
btrfs send /snapshots/@root-12.05.2018 | btrfs receive /data/snapshots/
```
`/data` is the mount point for an external hard disk. After some changes:
```
btrfs subvol snapshot -r / /snapshots/@root-20.05.2018
cd /snapshots
btrfs send -p @root-12.05.2018 @root-20.05.2018 | btrfs receive /data/snapshots/
```
At this point, I receive:
```
At subvol @root-20.05.2018
At snapshot @root-20.05.2018
ERROR: unlink etc/webmin/bind8/zone-names failed. No such file or directory
```
I was also thinking of something wrong or corrupted and tried different snapshot variations, but the incremental step fails **always**, just on another file. There is an `unlink` or `reflink` error thrown by btrfs.
The send/receive works perfectly if it is sent in full.
WTH?
My system:
```
# btrfs --version
btrfs-progs v4.7.3
```
```
# uname -r
4.9.61-odroidxu4
```
```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.10 (jessie)
Release:	8.10
Codename:	jessie
```
I know, it is not the newest, just the latest system shipped by Armbian. I could update, but seeing the report by @Nicop06, I'm afraid will not help.
Sorry again for bothering, I know it has nothing to do with btrbk. Just still don't believe it: btrfs incr send/receive does not work after 5 years?? How was it released then? Or I must be doing something terribly wrong...:()?'#
Any help or indication where to complain will be very appreciated. Many thanks.

P.S. How do you write multiline code blocks with this markdown?;)

**Comment by gkzsolt** [comment]
To add to the issue, send - receive works when the filesystems are on different servers. Also works (as I remember) if the volumes are on the same filesystem. But when different devices are mounted on one filesystem, it crashes.

**Comment by Nicop06** [comment]
Hi,

I do not think this is the same issue. I actually managed to solve mine, I will close this issue with the solution shortly. It was mostly due to data corruption, which may have been caused by a power shortage or kernel panic (I have many on my machines).

Your issue seems to be a common issue (I found some related problems when trying to solve mine) and looking on internet you might find some answers. Also, the #btrfs on freenode.

For block code, use ``` separator before and after your code.

**Comment by Jylpah** [comment]
@Nicop06 could you please post the solution here? I am having the same issue with Ubuntu 20.04 LTS & 5.4.10 kernel.

**Comment by Nicop06** [comment]
In my case that was due to a corruption. So I also had to copy all the files and format the disk, and then down from my backups. But I am not sure. I remember using a VM to debug the kernel, find  the corrupted volume and delete. And I probably ended up copying all the files with rsync and recreated the volume. But that was 2 years ago, and I didn't record what I did, and I don't really remember.

My conclusion was that btrfs is probably not a great idea to run if you are not ready to do this kind of debugging once it breaks. Ironically my backup server is also running btrfs. And I never had issues. But I highly recommend storing non-btrfs snapshot just in case, as file-system corruptions could be very bad and also propagate to backups.
